flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -221,9 +232,14 @@ public void checkpointState(
 		// We generally try to emit the checkpoint barrier as soon as possible to not affect downstream
 		// checkpoint alignments
 
+		if (lastCheckpointId >= metadata.getCheckpointId()) {
```

Comment:
For the below logic for checking abort request from RPC:

```
if (checkAndClearAbortedStatus(metadata.getCheckpointId())) {
			LOG.info("Checkpoint {} has been notified as aborted, would not trigger any checkpoint.", metadata.getCheckpointId());
			return;
}
```
Do we also need to call `channelStateWriter.abort` for RPC abort? If so, maybe we can integrate these two cases together for better handling the actions for aborting.

Reply:
Yes, I added it as a separate commit (it is just an optimization because we'll still get a "regular" abort call).

Reply:
In addition, it is a bit tough to understand the semantic of different cases now. The previous `lastCheckpointId` only reflects the already happened checkpoint, so wee can rely on `lastCheckpointId >= metadata.getCheckpointId()` to discard all the following smaller checkpoints based on the checkpoint id incremental guarantee.

Now the `lastCheckpointId` also reflects the canceled id from `CheckpointBarrierHandler`, so if the handler notifies to abort checkpoint 10, should we also need to ignore checkpoint 9. If so, it is also different with the abort case from RPC. If RPC notifies to abort checkpoint 10, the checkpoint 9 can still execute afterwards since we store the aborted id from RPC in the `abortedCheckpointIds` set.

Reply:
Yes, this is the right semantics and I agree it is more complex now. 
I described it in the comment for the aborted checkpoints field. 
However, I think this new complexity comes from the domain (or maybe design), and therefore can't be avoided in implementation.

Reply:
Yes, the current complexity somehow comes from coupling abort from handler with abort from RPC together, since `abortedCheckpointIds` will compare with `lastCheckpointId(aborted id from handler)` in some processes.

But indeed I can not think of an easy solution right now and also have not thought of critical problems ATM. So let's further discuss it if encountering problems future.